File: python/runtime/transport_envelope.py
# -*- coding: utf-8 -*-
"""Node/Python 間 transport の共通 envelope を扱うユーティリティ。"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Dict, Literal
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field

EnvelopeKind = Literal["command", "event", "status", "error"]
CURRENT_TRANSPORT_VERSION = "v1"


class TransportEnvelope(BaseModel):
    """contracts/transport-envelope.schema.json と対応する envelope。"""

    model_config = ConfigDict(extra="forbid")

    version: str = Field(min_length=2)
    trace_id: str = Field(min_length=1)
    run_id: str = Field(min_length=1)
    message_id: str = Field(min_length=1)
    timestamp: datetime
    source: str = Field(min_length=1)
    kind: EnvelopeKind
    name: str = Field(min_length=1)
    body: Dict[str, Any]
    auth: Dict[str, Any] | None = None


def make_transport_envelope(
    *,
    source: str,
    kind: EnvelopeKind,
    name: str,
    body: Dict[str, Any],
    trace_id: str | None = None,
    run_id: str | None = None,
    message_id: str | None = None,
) -> Dict[str, Any]:
    """送信時に利用する envelope を生成して dict で返す。"""

    envelope = TransportEnvelope(
        version=CURRENT_TRANSPORT_VERSION,
        trace_id=trace_id or uuid4().hex,
        run_id=run_id or uuid4().hex,
        message_id=message_id or uuid4().hex,
        timestamp=datetime.now(UTC),
        source=source,
        kind=kind,
        name=name,
        body=body,
    )
    return envelope.model_dump(mode="json")


def validate_transport_envelope(payload: Dict[str, Any]) -> TransportEnvelope:
    """受信 payload を検証し、型安全な envelope として返す。"""

    return TransportEnvelope.model_validate(payload)


__all__ = [
    "CURRENT_TRANSPORT_VERSION",
    "TransportEnvelope",
    "make_transport_envelope",
    "validate_transport_envelope",
]

File: node-bot/runtime/transportEnvelope.ts
import { randomUUID } from 'node:crypto';

import type { CommandPayload } from './types.js';

export const CURRENT_TRANSPORT_VERSION = 'v1';

export type EnvelopeKind = 'command' | 'event' | 'status' | 'error';

export interface TransportEnvelope {
  version: string;
  trace_id: string;
  run_id: string;
  message_id: string;
  timestamp: string;
  source: string;
  kind: EnvelopeKind;
  name: string;
  body: Record<string, unknown>;
  auth?: Record<string, unknown> | null;
}

export function buildEnvelope(params: {
  source: string;
  kind: EnvelopeKind;
  name: string;
  body: Record<string, unknown>;
  traceId?: string;
  runId?: string;
  messageId?: string;
}): TransportEnvelope {
  return {
    version: CURRENT_TRANSPORT_VERSION,
    trace_id: params.traceId ?? randomUUID(),
    run_id: params.runId ?? randomUUID(),
    message_id: params.messageId ?? randomUUID(),
    timestamp: new Date().toISOString(),
    source: params.source,
    kind: params.kind,
    name: params.name,
    body: params.body,
  };
}

export function validateEnvelope(input: unknown): TransportEnvelope | null {
  if (!input || typeof input !== 'object') {
    return null;
  }

  const maybe = input as Record<string, unknown>;
  const kind = maybe.kind;
  if (
    maybe.version !== CURRENT_TRANSPORT_VERSION ||
    typeof maybe.trace_id !== 'string' ||
    typeof maybe.run_id !== 'string' ||
    typeof maybe.message_id !== 'string' ||
    typeof maybe.timestamp !== 'string' ||
    typeof maybe.source !== 'string' ||
    (kind !== 'command' && kind !== 'event' && kind !== 'status' && kind !== 'error') ||
    typeof maybe.name !== 'string' ||
    !maybe.body ||
    typeof maybe.body !== 'object'
  ) {
    return null;
  }

  return maybe as unknown as TransportEnvelope;
}

export function adaptLegacyCommandPayload(input: unknown): TransportEnvelope | null {
  if (!input || typeof input !== 'object') {
    return null;
  }

  const payload = input as Partial<CommandPayload> & Record<string, unknown>;
  if (typeof payload.type !== 'string' || !payload.args || typeof payload.args !== 'object') {
    return null;
  }

  return buildEnvelope({
    source: 'legacy-python-agent',
    kind: 'command',
    name: payload.type,
    body: {
      type: payload.type,
      args: payload.args as Record<string, unknown>,
      ...(payload.meta && typeof payload.meta === 'object' ? { meta: payload.meta as Record<string, unknown> } : {}),
    },
  });
}
